### PR TITLE
PP-4790 Protect XRaySessionProfiler from NPE

### PR DESCRIPTION
**Pull request by georgeracu**
- Currently if the session is null, connector cannot connect to the database anymore

with @heathd

### CODE REVIEW

File: utils/src/main/java/uk/gov/pay/commons/utils/xray/XRaySessionProfiler.java [hunk 1]
```diff
@@ -37,36 +38,15 @@ public void endOperationProfile(String operationName, DatabaseQuery databaseQuer
...
-                hostname = new URI((new URI(metadata.getURL())).getSchemeSpecificPart()).getHost();
-                hostname = connection.getCatalog() + "@" + hostname;
-            } catch (URISyntaxException exception) {
-                logger.warn("Error parsing database host name.");
-            }
-        } catch (SQLException exception) {
-            logger.warn("Error getting database connection details.");
-        }
+        HashMap<String, Object> databaseMetadata = new HashMap<>();
+        String hostname = getDatabaseHostnameAndMetadata(databaseQuery, abstractSession, databaseMetadata);
```
Comment (alexbishop1) [inline utils/src/main/java/uk/gov/pay/commons/utils/xray/XRaySessionProfiler.java]: This method seems to get the hostname (defaulting to `"database"` if it can’t) and update the `databaseMetadata` map, which I found hard to follow.

If you take up the rest of my suggestions in the PR, we could replace these two lines with this:

```java
Optional<Connection> maybeConnection = Optional.ofNullable(abstractSession.getAccessor()).map(Accessor::getConnection);
Map<String, Object> databaseMetadata = maybeConnection.map(connection -> getMetadata(databaseQuery, connection)).orElseGet(HashMap::new);
String hostname = maybeConnection.flatMap(this::getDatabaseHostName).orElse("database");
```

File: utils/src/main/java/uk/gov/pay/commons/utils/xray/XRaySessionProfiler.java [hunk 2]
```diff
@@ -79,6 +59,51 @@ public Object profileExecutionOfQuery(DatabaseQuery databaseQuery, Record record
...
+            return hostname;
+        }
+        Connection connection = accessor.getConnection();
+        if (connection == null) {
+            return hostname;
+        }
+        updateMetadata(databaseQuery, databaseMetadata, connection);
+
+        return getDatabaseHostName(connection);
+    }
```
Comment (alexbishop1) [inline utils/src/main/java/uk/gov/pay/commons/utils/xray/XRaySessionProfiler.java]: If you follow the rest of my suggestions then this method can go away!

File: utils/src/main/java/uk/gov/pay/commons/utils/xray/XRaySessionProfiler.java [hunk 3]
```diff
@@ -79,6 +59,51 @@ public Object profileExecutionOfQuery(DatabaseQuery databaseQuery, Record record
...
+            databaseMetadata.put("user", metadata.getUserName());
+            databaseMetadata.put("driver_version", metadata.getDriverVersion());
+            databaseMetadata.put("database_type", metadata.getDatabaseProductName());
+            databaseMetadata.put("database_version", metadata.getDatabaseProductVersion());
+            databaseMetadata.put("preparation", databaseQuery.isCallQuery() ? "call" : "statement");
+            databaseMetadata.put("sanitized_query", StringUtils.isEmpty(databaseQuery.getSQLString()) ? "" : databaseQuery.getSQLString());
+        } catch (SQLException exception) {
+            logger.warn("Error getting database connection details.");
+        }
+    }
```
Comment (alexbishop1) [inline utils/src/main/java/uk/gov/pay/commons/utils/xray/XRaySessionProfiler.java]: I think this method would be better if it were `private void getMetadata(DatabaseQuery databaseQuery, Connection connection)` and returned a map (rather than mutating one):

```java
Map<String, Object> databaseMetadata = new HashMap<>();
try {
    // …
} catch (SQLException exception) {
    logger.warn("Error getting database connection details.");
}
return databaseMetadata;
```

File: utils/src/main/java/uk/gov/pay/commons/utils/xray/XRaySessionProfiler.java [hunk 4]
```diff
@@ -79,6 +59,51 @@ public Object profileExecutionOfQuery(DatabaseQuery databaseQuery, Record record
...
+            DatabaseMetaData metadata = connection.getMetaData();
+            hostname = new URI((new URI(metadata.getURL())).getSchemeSpecificPart()).getHost();
+            return connection.getCatalog() + "@" + hostname;
+        } catch (URISyntaxException exception) {
+            logger.warn("Error parsing database host name.");
+        } catch (SQLException exception) {
+            logger.warn("Error getting database connection details.");
+        }
+        return hostname;
+    }
```
Comment (alexbishop1) [inline utils/src/main/java/uk/gov/pay/commons/utils/xray/XRaySessionProfiler.java]: Rather than returning a default `"database"` value, this method could return an `Optional<String>` and allow the caller to decide what the default is.

File: utils/src/main/java/uk/gov/pay/commons/utils/xray/XRaySessionProfiler.java (hunk 3 above)
Comment (rhowe-gds) [reply utils/src/main/java/uk/gov/pay/commons/utils/xray/XRaySessionProfiler.java]: I'd also like to see the exception being passed to the logger

File: utils/src/main/java/uk/gov/pay/commons/utils/xray/XRaySessionProfiler.java [hunk 5]
```diff
@@ -36,37 +40,23 @@ public void endOperationProfile(String operationName, DatabaseQuery databaseQuer
...
-            additionalParams.put("database_version", metadata.getDatabaseProductVersion());
-            additionalParams.put("preparation", databaseQuery.isCallQuery() ? "call" : "statement");
-            additionalParams.put("sanitized_query", StringUtils.isEmpty(databaseQuery.getSQLString()) ? "" : databaseQuery.getSQLString());
-
-            try {
-                hostname = new URI((new URI(metadata.getURL())).getSchemeSpecificPart()).getHost();
-                hostname = connection.getCatalog() + "@" + hostname;
-            } catch (URISyntaxException exception) {
-                logger.warn("Error parsing database host name.");
+        if (recorder.getCurrentSegmentOptional().isPresent()) {
```
Comment (alexbishop1) [inline utils/src/main/java/uk/gov/pay/commons/utils/xray/XRaySessionProfiler.java]: I am so looking forward to when we can use `ifPresentOrElse(…)`!

File: utils/src/main/java/uk/gov/pay/commons/utils/xray/XRaySessionProfiler.java [hunk 6]
```diff
@@ -36,37 +40,23 @@ public void endOperationProfile(String operationName, DatabaseQuery databaseQuer
...
+            final Optional<Connection> maybeConnection = Optional.ofNullable(abstractSession.getAccessor()).map(Accessor::getConnection);
+            Map<String, Object> databaseMetadata = new HashMap<>();
+            String hostname = "database";
+            if (maybeConnection.isPresent()) {
+                final Connection connection = maybeConnection.get();
+                databaseMetadata = getMetadata(databaseQuery, connection);
+                hostname = getDatabaseHostName(connection);
+            } else {
+                logger.warn("There is no database connection available");
```
Comment (alexbishop1) [inline utils/src/main/java/uk/gov/pay/commons/utils/xray/XRaySessionProfiler.java]: Should this (and possibly other logging in this class) be `error`? Something’s gone pretty wrong if we don’t have a database connection!

File: utils/src/main/java/uk/gov/pay/commons/utils/xray/XRaySessionProfiler.java [hunk 7]
```diff
@@ -36,37 +40,23 @@ public void endOperationProfile(String operationName, DatabaseQuery databaseQuer
...
-            logger.warn("Error getting database connection details.");
-        }
 
-        if (recorder.getCurrentSegmentOptional().isPresent()) {
             Subsegment subsegment = recorder.beginSubsegment(hostname);
             subsegment.putMetadata("monitor_name", databaseQuery.getMonitorName());
             subsegment.putMetadata("calling_class", databaseQuery.getClass().getSimpleName());
             subsegment.setNamespace(Namespace.REMOTE.toString());
-            subsegment.putAllSql(additionalParams);
+            subsegment.putAllSql(databaseMetadata);
```
Comment (alexbishop1) [inline utils/src/main/java/uk/gov/pay/commons/utils/xray/XRaySessionProfiler.java]: This looks a lot nicer but it’s a shame that things like the `"database"` default and the empty database metadata map appear in two places.

We could do something like this?

```java
if (recorder.getCurrentSegmentOptional().isPresent()) {
    final Optional<Connection> maybeConnection = Optional.ofNullable(abstractSession.getAccessor()).map(Accessor::getConnection);
    
    if (!maybeConnection.isPresent()) {
        logger.warn("There is no database connection available");
    }

    Map<String, Object> databaseMetadata = maybeConnection.map(connection -> getMetadata(databaseQuery, connection)).orElseGet(HashMap::new);
    String hostname = maybeConnection.flatMap(this::getDatabaseHostName).orElse("database");

    Subsegment subsegment = recorder.beginSubsegment(hostname);
    subsegment.putMetadata("monitor_name", databaseQuery.getMonitorName());
    subsegment.putMetadata("calling_class", databaseQuery.getClass().getSimpleName());
    subsegment.setNamespace(Namespace.REMOTE.toString());
    subsegment.putAllSql(databaseMetadata);

    try {
        return abstractSession.internalExecuteQuery(databaseQuery, (AbstractRecord) record);
    } finally {
        subsegment.end();
    }
} else {
    logger.info("No segment found, executing query without recording subsegment data.");
    return abstractSession.internalExecuteQuery(databaseQuery, (AbstractRecord) record);
}
```

This would require following the earlier suggestion to make `getDatabaseHostName(Connection)` return an `Optional<String>`.

File: utils/src/main/java/uk/gov/pay/commons/utils/xray/XRaySessionProfiler.java [hunk 8]
```diff
@@ -79,6 +69,37 @@ public Object profileExecutionOfQuery(DatabaseQuery databaseQuery, Record record
...
+            databaseMetadata.put("driver_version", metadata.getDriverVersion());
+            databaseMetadata.put("database_type", metadata.getDatabaseProductName());
+            databaseMetadata.put("database_version", metadata.getDatabaseProductVersion());
+            databaseMetadata.put("preparation", databaseQuery.isCallQuery() ? "call" : "statement");
+            databaseMetadata.put("sanitized_query", StringUtils.isEmpty(databaseQuery.getSQLString()) ? "" : databaseQuery.getSQLString());
+            return databaseMetadata;
+        } catch (SQLException exception) {
+            logger.warn("Error getting database connection details.");
+        }
+        return Collections.emptyMap();
```
Comment (alexbishop1) [inline utils/src/main/java/uk/gov/pay/commons/utils/xray/XRaySessionProfiler.java]: I assume it’s safe for this to be an immutable map?
